### ISSUE TITLE
Change log: Updates to display of disaggregation values

### ISSUE THREAD

**Issue by jennyhavah** [mercycorps/TolaActivity#1520]
This work is peeled off of #1481 

# Issues
Here is how the change log appears if you add disaggregated values to some categories and not others.

![image](https://user-images.githubusercontent.com/33670923/56920857-31c39880-6a79-11e9-8f44-142d59b3a6ca.png)

We include categories whose values have not changed, with "N/A" not making much sense in this context. It would be better to not include categories whose values have not changed.

From Ken: The N/A -> empty string above is the result of an existing implementation detail of how we store disaggregation values. Basically we store nothing, until 1 value is added, at which point we create entries for every possible desegregation value in the form.

In the above example, the N/A entry on the left represents non-existence, where the right represents existence with an empty string value that exists in the DB. From an annoyingly pedantic standpoint, it did change.

From Jenny: From a user perspective, "N/A" doesn't make sense. Also, as a general rule, we don't want to be displaying values in the change log that did not change -- _from a user perspective._

# Acceptance criteria

In the change log, the following things are true about changes to Results
- [x] 1. We only display disaggregated values in the change log when they were modified _by the end user_ -- meaning it went from empty to filled, from filled to empty, or from one filled value to a different filled value.
- [x] 2. I do not see "N/A" in the change log at any time... If we're trying to communicate that a field is empty, then we display the disaggregation category followed by a colon and empty space.
- [x] 3. When both disaggregated and total values have changed, display the total value first, followed by the disaggregated values.
- [x] 4. The system ID should not appear in the change log
- [x] 5. Sites should have a capital "S"
- [x] 6. Match row labels to the labels on the form that the user sees
- [x] 7. Ensure disaggregation labels are being ordered alphabetically by name and then by the label sort order saved in the db
- [x] 8. Change should not be logged when there is no change.  The value currently triggers a change, even if there isn't one (e.g. it compares 23 to 23.00)
- [x] 9. Change type should sentence case
- [x] 10. Associate the "Disaggregated values changed" header and its data more closely
- [x] ~11. When a disaggregation is nulled in the front end, it appears in the change log as 0.  Instead the value should be an empty string.~ Now a separate ticket,  #2095.
- [x] 12. Change label text of "Disaggregated values changed" to "Disaggregated values"

In the change log, the following things are true about changes to Indicators
- [x] 13.  If it's the indicator name that's changed, put a "heading" into the previous and new cells, similar to what's done with other types of changes. [edit:] The heading should be the same as the one used on the program page, e.g. Indicator a, Indicator 1.2c, etc...  Fall back to "Indicator" for older change log entries.

Changes to RF Level tracking
- [x] 14. For changes to the name of an RF level, use the tier and ontology as the label. For instance Outcome 1.2.  Fall back to "Result level" for older change log entries.

## Housekeeping (where applicable)
- [ ] Unit test/s written 
- [ ] i18n strings included
- [ ] Is accessible to people using screen readers

**Comment by camgnostic** [comment]
Tagged "HIGH" because the change log code has a lot of "magic" and unintended consequences in it.  Might turn out to be easier.

**Comment by sanjuroj** [comment]
@jennyhavah, a couple of questions.  

On item #2, do you the total to appear if only the disaggregations have changed?  Or should just the updated disaggregations appear in that case.  

Related to item #11, I noticed that when a disaggregation is nulled out, it's actually set to 0 rather than null.  There's a difference in the way that shows up on the front end (as a 0 rather than an empty string).  I'm guessing it should be set to null rather than 0 for consistency with other unused disaggregation categories, but I wanted to get your input.  Maybe the current behavior is correct?

**Comment by jennyhavah** [comment]
> On item #2, do you the total to appear if only the disaggregations have changed? Or should just the updated disaggregations appear in that case.

@sanjuroj I'd like to see the total (and to be clear, by that I mean display the "Actual value" field) be displayed whenever disaggregations have changed. In some cases, the actual value will change based on the updated disaggs. Even if that is not the case, it's important context to know that the total did NOT change.

**Comment by jennyhavah** [comment]
> Related to item #11, I noticed that when a disaggregation is nulled out, it's actually set to 0 rather than null. There's a difference in the way that shows up on the front end (as a 0 rather than an empty string). I'm guessing it should be set to null rather than 0 for consistency with other unused disaggregation categories, but I wanted to get your input. Maybe the current behavior is correct?

@sanjuroj My instinct is that it's important to have a concept of a "null" value for disaggregated values -- that it's problematic to auto-populate with a zero. A result of zero is potentially a meaningful result, not always equivalent to "no data for this category." Carly agrees, and suggested that we use the same long dash we use in IPTTs and the program page, in this scenario. (This also means we don't have to require that users manually enter a bajillion zeroes in order to avoid an ambiguous mix of empties and zeroes in a single column of data.)

This back end change is outside the scope of this change log ticket, so I filed a new one and assigned it to @camgnostic -- #2095

**Comment by sanjuroj** [comment]
@jennyhavah, related to making the change log labels more similar to the result form labels, 2 questions.

1.  On the Result for the label is "Site" while in the change log it's "Sites".  This is a multi-select field, so there could be one or more sites associated with the result.  Preference for "Sites" or "Site" (or something else)?
2.  'Measure against target" feels a bit cryptic to put in the change log.  The current label is "Target", is this ok?  Also, the period name is given without providing the dates, so it's not obvious what date range is being affected.  It might be better to provide the date range instead, however, anything to do with dates feels potentially non-trivial to me.  Any thoughts?

**Comment by jennyhavah** [comment]
@sanjuroj 

> On the Result for the label is "Site" while in the change log it's "Sites". This is a multi-select field, so there could be one or more sites associated with the result. Preference for "Sites" or "Site" (or something else)?

As you say, it's a multi-select, so let's make it "Sites" both on the result form label and in the change log. (Assuming it's an easy change in both places, let me know if that's not the case.)

> 'Measure against target" feels a bit cryptic to put in the change log. The current label is "Target", is this ok? Also, the period name is given without providing the dates, so it's not obvious what date range is being affected. It might be better to provide the date range instead, however, anything to do with dates feels potentially non-trivial to me. Any thoughts?

I would prefer to see "Measure against target" to be consistent with the label in the form. "Target" by itself suggests a value, and the field tells you which target period we're measuring against.

In a similar vein, I would prefer that "Date" be replaced with "Result date" and that this field be placed before the "Measure against target" field -- on the general principle that the field order in the change log ought to mirror the field order in the form whose changes we're recording.

**Comment by sanjuroj** [comment]
@jennyhavah Since we're talking overall sorting, what do you want to do in the scenario where the result value hasn't changed, there are disaggregation changes, and the Sites field has changed?  In this case the order as dictated by the result form would put data in between the result value and the disaggregations.  Would you like to preserve the form order?  Or do you still want the result value to be just above the changed disaggregations.  And what if the result value has changed (i.e. result value, sites, and disaggs all changed)?

**Comment by jennyhavah** [comment]
@sanjuroj 

> Since we're talking overall sorting, what do you want to do in the scenario where the result value hasn't changed, there are disaggregation changes, and the Sites field has changed? In this case the order as dictated by the result form would put data in between the result value and the disaggregations. Would you like to preserve the form order? Or do you still want the result value to be just above the changed disaggregations. And what if the result value has changed (i.e. result value, sites, and disaggs all changed)?

I'm having trouble seeing why Sites would appear in between the result value and the disaggregation. Why would that happen? In both the current implementation and the Goat Tree design, Sites comes after disaggregations.

If the "Actual value" field has not changed, I don't think we need to display it, regardless of whether the disaggregated field values have changed.

If the "Actual value" field HAS changed, then I think it should come first, before the disaggregated value changes.

In general, I'm trying to keep the change log rules as clear and simple as possible -- only using the same labels, only showing changed fields, and mimicking the result form order. Definitely open to changing the rules it if users find it confusing. But I figure this is a good place to start, an improvement over today's labeling and ordering that comes across as somewhat arbitrary.

**Comment by sanjuroj** [comment]
@axoplasm Assigning to you for one side of the design sandwich.  Please take a look at the change log entries to see if there is a way to style them better so the diffs are properly grouped together and it's clear what is a header and what is data.  The "Community Investment Trust" program has some change log entries with disaggregations.

**Comment by axoplasm** [comment]
> 2. I do not see "N/A" in the change log at any time... If we're trying to communicate that a field is empty, then we display the disaggregation category followed by a colon and empty space.

IANOU but this looks broken to me. I’d expect to see something to indicate `null`, like an em-dash (my preference), N/A, or “none.” It also makes the changed values vanish in a sea of empty values. 

Everything but an em-dash looks noisy; we could probably style this away by making the placeholder light gray.

also: the data are ordered in a way that takes a little parsing. “disaggregated values” is really subheader for “Actual value,” but a header for the disaggregation name, which is itself a header for the data. It *should* look something like the screenshot, but this is with a lot of fiddling. This runs into a couple semi-global problems:

— we’ve kind of hacked `<h*>` headers so that they don’t form a consistent hierarchy 
— dang near everything in this display is bold now! (If everyone is shouting, you can’t hear anyone)

![image](https://user-images.githubusercontent.com/169060/73206694-3b595a00-40f8-11ea-8b91-93e5da814972.png)

**Comment by jennyhavah** [comment]
> IANOU but this looks broken to me. I’d expect to see something to indicate null, like an em-dash (my preference), N/A, or “none.” It also makes the changed values vanish in a sea of empty values.

I agree that using the em-dash makes sense any case the user has not entered a value. This is consistent with the pattern we're using on the Program Page and IPTT.

> Everything but an em-dash looks noisy; we could probably style this away by making null rows gray.

I have some concerns about using gray in addition to the em-dash. First, that the gray could be misinterpreted as "not applicable" or something other than "no value was entered." On top of that, there's a difference between zero and "no value entered" -- both are valid and both mean different things. (Zero could mean "we tried this and it had zero impact." I think the gray adds ambiguity and complexity. Also, it will be very common for alternating disaggregation categories to have values or not have values, so you could end up with an irregular stripey effect if we only apply the gray to categories that don't have values. It would add busy-ness rather than reducing it.

> also: the data are ordered in a way that takes a little parsing. “disaggregated values” is really subheader for “Actual value,” but a header for the disaggregation name, which is itself a header for the data. 

I don't think the sub-header "Disaggregated values" adds anything and would like to suggest we take it out entirely, leaving the disaggregation types (SADD, etc.) as the sub-headers.

> It should look something like the screenshot, but this is with a lot of fiddling. This runs into a couple semi-global problems:
> — we’ve kind of hacked <h*> headers so that they don’t form a consistent hierarchy
> — dang near everything in this display is bold now! (If everyone is shouting, you can’t hear anyone)

Paul, what do you think about not having the disagg categories be bold? So it'd look more like this:

**Actual value:** 54

**Sex and Age Disaggregated Data (SADD)**
Male 0-5: --
Female 0-5: --

**Comment by axoplasm** [comment]
I like that better than all the shouty bolds. We could also highlight changed values with color:

![image](https://user-images.githubusercontent.com/169060/73218728-63a08300-410f-11ea-8e01-14371930c1f9.png)

**Comment by jennyhavah** [comment]
As Sanjuro just pointed out in Slack, we only show all form fields (including all disagg categories) when the result is first added. After that, we only show form fields (including disagg values) that changed. This means that the blue would lose its meaning, because every disagg change log entry would have blue numbers.

**Comment by axoplasm** [comment]
OK, let’s leave out the primary highlight for now.

**Comment by jennyhavah** [comment]
@axoplasm Not seeing em-dashes when a field has a null value.

**Comment by jennyhavah** [comment]
@sanjuroj When editing the disaggregations of  _saved results_, I am seeing disagg categories show up in the change log that were not modified.

In this example on QA Program - Jenny, I made changes to Chocolate chip and Peanut butter, but not to Rugelach.

![image](https://user-images.githubusercontent.com/33670923/73499003-bf6b4600-4373-11ea-8118-83bede271d06.png)

**Comment by sanjuroj** [comment]
@jennyhavah Do you want the em-dashes to appear in place of N/A in the Result level and Indicator columns when those values are blank?  For instance, if you change  a result level, there is currently an N/A in the Indicator column.

**Comment by sanjuroj** [comment]
@jennyhavah I'm trying to understand why that blank entry is appearing for Rugelach.  Is it possible that you added that the Rugelach category after you initially created the result?  This might trick the system into showing the difference.

**Comment by jennyhavah** [comment]
> @jennyhavah Do you want the em-dashes to appear in place of N/A in the Result level and Indicator columns when those values are blank? For instance, if you change a result level, there is currently an N/A in the Indicator column.

This sounded unexpected to me, so I tried changing a result level for a saved indicator. I don't see any N/A in the indicator or result level columns, and I wouldn't expect to. Here's what I'm seeing:

![image](https://user-images.githubusercontent.com/33670923/73709416-de404400-46b5-11ea-819d-c5d27a63fa45.png)

Are we talking about the same thing?

**Comment by jennyhavah** [comment]
> @jennyhavah I'm trying to understand why that blank entry is appearing for Rugelach. Is it possible that you added that the Rugelach category after you initially created the result? This might trick the system into showing the difference.

@sanjuroj  Yes, that's exactly what happened! I was able to replicate it like so:

![image](https://user-images.githubusercontent.com/33670923/73710261-5ad42200-46b8-11ea-8e6f-7ee29890d7cc.png)

I must have been juggling between testing this ticket and another one related to categories. Sorry for sending you on a wild goose chase!

Just to be clear, I'd still like em dashes to appear when a disagg category has no value entered... I can think of two ways this might happen:

1. New result, all disagg categories listed but not all have values.
2. Edited result: A value is removed from a category and added to a different category. The first category is now empty.

**Comment by sanjuroj** [comment]
> This sounded unexpected to me, so I tried changing a result level for a saved indicator. I don't see any N/A in the indicator or result level columns, and I wouldn't expect to. Here's what I'm seeing...

@jennyhavah I'll have to look on my work computer tomorrow to see how I did this with result level.  But you can certainly get N/A's if you change the program dates.

**Comment by jennyhavah** [comment]
@sanjuroj Aha, yes, I do see it for the program date change:

![image](https://user-images.githubusercontent.com/33670923/73711569-295d5580-46bc-11ea-81bb-e5e661b61259.png)

There's clearly a lot of inconsistency in how we handle null values in the change log. For Goat Tree -- and in general -- I'm only concerned about consistently showing em dashes when the field is numeric. This is because you're often viewing many numeric fields close together -- like the IPTT or the Program page or the change log -- and so the variability really increases cognitive load. 

Which is to say, I don't think we need to expand the scope of this ticket to other types of changes besides results.

**Comment by sanjuroj** [comment]
@jennyhavah It still feels to me like Stroopwafel should not be showing up in your image above.  Want me to look into that?  

And I finished the em dashes, just haven't done a PR yet, waiting on some of these other things we're talking about.

Also, I just figured out why you didn't get an NA in indicator.  You changed the result level of an indicator (i.e. you switched from one level to another).  Try changing the result lavel itself (like the name of the level).

**Comment by jennyhavah** [comment]
> Also, I just figured out why you didn't get an NA in indicator. You changed the result level of an indicator (i.e. you switched from one level to another). Try changing the result lavel itself (like the name of the level).

You mean like create a custom template and then change the level? I actually don't think "N/A" is necessarily wrong. "This field is not relevant to the change we're talking about" is a bit different than "This disaggregation field is totes legit, there's just no value for it."

I'm not inclined to tackle this right now -- or at least, not as part of this ticket.

**Comment by jennyhavah** [comment]
> @jennyhavah It still feels to me like Stroopwafel should not be showing up in your image above. Want me to look into that?

@sanjuroj True, it shouldn't appear. It's an extreme edge case, though, so I would time-box it. If it's looking like it'll take more than an hour to resolve, let's break it off into a Poblano ticket.

**Comment by sanjuroj** [comment]
> I actually don't think "N/A" is necessarily wrong.
I agree, changing the name of a level shouldn't be associated with a particular indicator or set of indicators.  I'll explain the rest tomorrow.

**Comment by jennyhavah** [comment]
@sanjuroj Found some odd codes showing up when I edit a result:

![image](https://user-images.githubusercontent.com/33670923/73970789-9cd8b000-48d2-11ea-9c3d-824b0412c31f.png)

**Comment by jennyhavah** [comment]
@sanjuroj Looks like some of Paul's formatting might have reverted? Looks like some vertical spacing is gone between Actual value and the first disagg type, and before/after disagg types. See this screenshot for what it used to look like: https://github.com/mercycorps/TolaActivity/issues/1520#issuecomment-581721910

**Comment by jennyhavah** [comment]
@sanjuroj I just tried retesting the condition where I add a new disaggregation category and then edit a saved result. All I did was edit Peanut butter. But I'm seeing _all_ categories now, instead of just the one that I added -- Black and white. This is in QA - Jenny.

![image](https://user-images.githubusercontent.com/33670923/73971325-87b05100-48d3-11ea-9aa4-71ba9a296403.png)

**Comment by jennyhavah** [comment]
@sanjuroj Steps taken on QA:

1. Created a new country disaggregation for Afghanistan called "Jenny's test disaggregation" and added three categories to it.

2. Assigned new disaggregation to indicator I.1.1. from the Employable, Skilled and Inclusive Workforce (ESIF) program.

3. Added a result disaggregated by "Jenny's..."

4. Returned to Admin and added a new category to my disaggregation: Pandas.

5. Edited the previous result but NOT disaggregating by Pandas.

6. Visited change log. Here's what I'm seeing:

![image](https://user-images.githubusercontent.com/33670923/74182691-5f3c9580-4bf8-11ea-9212-eef97f571dd3.png)

Again, I didn't disaggregate that result by Pandas, just edited the numbers for the categories I'd already used.

**Comment by sanjuroj** [comment]
@jennyhavah I think this example you just posted matches the initial post about a week ago that we said I should time box.  This actually ties into the bigger picture ticket I just filed in Poblano about refactoring the diff functionality for the audit log.  (#2198).  I think fixing this is outside of the time box that we set for it.

However, this example also seems different than your message from just a few days ago where all of the disaggregation categories were showing up when you changed a result, even though only one category value had changed.  Is that something you're able to replicate?  Or did I just miss something in your most recent example?

**Comment by jennyhavah** [comment]
> However, this example also seems different than your message from just a few days ago where all of the disaggregation categories were showing up when you changed a result, even though only one category value had changed. Is that something you're able to replicate? Or did I just miss something in your most recent example?

@sanjuroj I did a bunch of testing on this and forgot to write it up! I was not able to replicate the issue of all categories showing up when I only edited a few.. It's entirely possible I got confused, thinking I was editing an existing disaggregation but was really adding a new (parallel) disaggregation. So let's consider that issue resolved, and I'll keep an eye on it as I do further testing...
